pkg/driver/file/device_manager_file.go

```diff
@@ -393,6 +395,25 @@ func (d *DeviceManager) DeviceCheckCert(cert *x509.Certificate) (*uuid.UUID, err
 	return nil, nil
 }
 
+// DeviceCheckCertHash see if a particular certificate hash is a valid registered device certificate
+func (d *DeviceManager) DeviceCheckCertHash(hash []byte) (*uuid.UUID, error) {
+	if hash == nil {
+		return nil, fmt.Errorf("invalid empty hash")
+	}
+	// refresh certs from filesystem, if needed - includes checking if necessary based on timer
+	err := d.refreshCache()
+	if err != nil {
+		return nil, fmt.Errorf("unable to refresh certs from Redis: %v", err)
+	}
+	for k, u := range d.deviceCerts {
+		s := sha256.Sum256([]byte(k))
```

Comment:
It looks like this change materially reduces the security, but I probably misunderstood.

Beforehand, we actually relied on the mTLS protocol to ensure that the device cert presented by the client was validated (i.e. it actually was able to do the encryption and signature that are part of mTLS). We then were able to rely on that to just get the UUID from the cert. [This line](https://github.com/lf-edge/adam/pull/100/files#diff-104a11712684e520f0affa95e6a053faeb713306a10f69ee99b9729cc9dd1996R276) got the cert from the go http reader, which had already done the mutual validation.

From what I can tell here, we are just taking the hash of the cert in the payload. Has anything validated that the device actually has the cert in question? It looks like the authcontainer just has the hash in it, which leaves it wide open to MITM attacks and replay attacks.

Reply:
Great point, thank you! Seems, I miss validation of AuthContainer`s SignatureHash against registered certificates.
Added into validateAuthContainerAndRecord.